### PR TITLE
Primary nodes should surive primary stepdowns

### PR DESCRIPTION
**Pull request by rozza**
JAVA-3075

https://evergreen.mongodb.com/version/5cd588b6c9ec441dea3e69b0##1

### CODE REVIEW

**Comment by rozza** [comment]
Missed the async tests - will update soon

File: driver-core/src/main/com/mongodb/internal/connection/DefaultServer.java [hunk 1]
```diff
@@ -156,11 +162,19 @@ ConnectionPool getConnectionPool() {
...
-            || t instanceof MongoNodeIsRecoveringException) {
+        if ((t instanceof MongoSocketException && !(t instanceof MongoSocketReadTimeoutException))) {
             invalidate();
+        } else if (t instanceof MongoNotPrimaryException || t instanceof MongoNodeIsRecoveringException) {
+            if (description.getMaxWireVersion() < FOUR_DOT_TWO_WIRE_VERSION) {
+                invalidate();
+            } else if (SHUTDOWN_CODES.contains(((MongoCommandException) t).getErrorCode())) {
+                invalidate();
+            } else {
+                serverStateListener.stateChanged(new ChangeEvent<ServerDescription>(description, ServerDescription.builder()
```
[the remark this answers is not in the document]
Comment (rozza) [reply driver-core/src/main/com/mongodb/internal/connection/DefaultServer.java]: Its following the [SDAM spec](https://github.com/mongodb/specifications/blob/master/source/server-discovery-and-monitoring/server-discovery-and-monitoring.rst#id79): 

> When the client sees a "not master" or "node is recovering" error it MUST replace the server's description with a default ServerDescription of type Unknown. It MUST store useful information in the new ServerDescription's error field, including the error message from the server.

File: driver-core/src/main/com/mongodb/internal/connection/DefaultServer.java [hunk 2]
```diff
@@ -156,11 +162,19 @@ ConnectionPool getConnectionPool() {
...
+                        .state(CONNECTED).type(ServerType.UNKNOWN).address(serverId.getAddress()).exception(t).build()));
```
[the remark this answers is not in the document]
Comment (rozza) [reply driver-core/src/main/com/mongodb/internal/connection/DefaultServer.java]: Can you repost the link? Or the lines of code - I can't find that diff

[the remark this answers is not in the document]
Comment (jyemin) [reply driver-core/src/main/com/mongodb/internal/connection/DefaultServer.java]: https://github.com/mongodb/specifications/commit/cb1c5fea75a3d44f1368b073ea91aa00ba3d338c#diff-937efe3124be43f659bef57d71d9a112

It's in the `source/server-discovery-and-monitoring/server-discovery-and-monitoring.rst` file

File: driver-core/src/main/com/mongodb/internal/connection/DefaultServer.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (jyemin) [reply driver-core/src/main/com/mongodb/internal/connection/DefaultServer.java]: Ah, ok.  I think it makes more sense to use `.state(CONNECTING)` in that case.  It shouldn't even be allowed to mix ServerType UNKNOWN with state of CONNECTED.

File: driver-core/src/main/com/mongodb/internal/connection/DefaultServer.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (rozza) [reply driver-core/src/main/com/mongodb/internal/connection/DefaultServer.java]: Ack - thanks

File: driver-core/src/main/com/mongodb/internal/connection/DefaultServer.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (rozza) [reply driver-core/src/main/com/mongodb/internal/connection/DefaultServer.java]: Done

**Comment by rozza** [comment]
https://evergreen.mongodb.com/version/5cdad6cad6d80a7884bdc211##1

File: driver-core/src/main/com/mongodb/internal/connection/DefaultServer.java [hunk 3]
```diff
@@ -132,6 +138,24 @@ public void invalidate() {
...
+    @Override
+    public void invalidate(final Throwable t) {
+        if ((t instanceof MongoSocketException && !(t instanceof MongoSocketReadTimeoutException))) {
+            invalidate();
+        } else if (t instanceof MongoNotPrimaryException || t instanceof MongoNodeIsRecoveringException) {
+            if (description.getMaxWireVersion() < FOUR_DOT_TWO_WIRE_VERSION) {
+                invalidate();
+            } else if (SHUTDOWN_CODES.contains(((MongoCommandException) t).getErrorCode())) {
+                invalidate();
+            } else {
```
Comment (jyemin) [inline driver-core/src/main/com/mongodb/internal/connection/DefaultServer.java]: Let's protect this block (or the whole method) with an isClosed() check, to match the other invalidate method.

Comment (rozza) [reply driver-core/src/main/com/mongodb/internal/connection/DefaultServer.java]: Ack - will do

**Comment by rozza** [comment]
> Any hypothesis on why this fixes the transaction tests?

Not yet but I want to take some time to understand why it does

**Pull request closed without merging**